src/Symfony/Component/Finder/Iterator/VcsIgnoredFilterIterator.php

```diff
@@ -37,25 +42,52 @@ public function accept(): bool
         $file = $this->current();
 
         $fileRealPath = $this->normalizePath($file->getRealPath());
-        if ($file->isDir() && !str_ends_with($fileRealPath, '/')) {
+
+        return !$this->isIgnored($fileRealPath);
+    }
+
+    private function isIgnored(string $fileRealPath): bool
+    {
+        if (is_dir($fileRealPath) && !str_ends_with($fileRealPath, '/')) {
             $fileRealPath .= '/';
         }
 
+        if (isset($this->ignoredPathsCache[$fileRealPath])) {
+            return $this->ignoredPathsCache[$fileRealPath];
+        }
+
+        $ignored = false;
+
         foreach ($this->parentsDirectoryDownward($fileRealPath) as $parentDirectory) {
+            if ($this->isIgnored($parentDirectory)) {
+                $ignored = true;
+
+                // rules in ignored directories are ignored, no need to check further.
+                break;
+            }
+
             $fileRelativePath = substr($fileRealPath, \strlen($parentDirectory) + 1);
 
-            $regex = $this->readGitignoreFile("{$parentDirectory}/.gitignore");
+            $regexps = $this->readGitignoreFile("{$parentDirectory}/.gitignore");
 
-            if (null !== $regex && preg_match($regex, $fileRelativePath)) {
-                return false;
+            if (null === $regexps) {
```

Comment:
can be merged with the previous line `if (null === $regexps = ...`

Reply:
Done.